Java spring cinema booking quickstart (#1916)
lightbend/kalix-jvm-sdk: samples/java-spring-cinema-booking-choreography/README.md

> # Cinema Booking sample

Reviewer: There are a lot of headings and nested headings. I suggest adding a table of contents to make it easier to understand the structure of the content and move from section to section.

```suggestion
# Cinema Booking sample

## Table of Contents
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
  - [Prerequisites](#prerequisites)
  - [Kalix SDK](#kalix-sdk)
- [Designing Kalix Services](#designing-kalix-services)
  - [Use case definition](#use-case-definition)
  - [Event storming](#event-storming)
    - [Entities / Aggregate Roots](#entities--aggregate-roots)
    - [Commands & Domain Events](#commands--domain-events)
  - [Bounded contexts](#bounded-contexts)
    - [Show](#show)
    - [Wallet](#wallet)
  - [Context map](#context-map)
  - [Non-functional requirements](#non-functional-requirements)
  - [Kalix components](#kalix-components)
    - [Show](#show-1)
    - [Wallet](#wallet-1)
    - [Context map](#context-map-1)
- [Kickstart a Kalix development project](#kickstart-a-kalix-development-project)
  - [Maven archetype](#maven-archetype)
  - [Import the generated project into your IDE/editor](#import-the-generated-project-into-your-ideeditor)
  - [Adding additional, non-mandatory dependencies](#adding-additional-non-mandatory-dependencies)
  - [Java 21](#java-21)
- [Develop Kalix components for the "Show" bounded context](#develop-kalix-components-for-the-show-bounded-context)
  - [Describe Data structures](#describe-data-structures)
    - [API data structures](#api-data-structures)
    - [Domain Events data structure](#domain-events-data-structure)
    - [Show entity state data structure and business domain logic encapsulation](#show-entity-state-data-structure-and-business-domain-logic-encapsulation)
  - [Develop the "Show" Entity](#develop-the-show-entity)
  - [Show Entity unit test](#show-entity-unit-test)
    - [Run unit test](#run-unit-test)
  - [Run & test locally](#run--test-locally)
    - [Run](#run)
    - [Test](#test)
  - [Developing the "Shows By Availability" View](#developing-the-shows-by-availability-view)
  - [View integration test](#view-integration-test)
    - [Run integration test](#run-integration-test)
  - [Developing the "Reservation" Entity](#developing-the-reservation-entity)
  - [Developing the `FoldShowEventsToReservation` Kalix Action](#developing-the-foldshoweventstoreservation-kalix-action)
  - [Entity and Action interaction integration test](#entity-and-action-interaction-integration-test)
    - [Run integration test](#run-integration-test-1)
- [Developing Kalix components for "Wallet" bounded context](#developing-kalix-components-for-wallet-bounded-context)
  - [Describing Data structures](#describing-data-structures)
    - [API data structures](#api-data-structures-1)
    - [Domain Events data structure](#domain-events-data-structure-1)
    - [Wallet entity state data structure and business domain logic encapsulation](#wallet-entity-state-data-structure-and-business-domain-logic-encapsulation)
  - [Developing the "Wallet" Entity](#developing-the-wallet-entity)
  - [Wallet Entity unit test](#wallet-entity-unit-test)
    - [Run unit test](#run-unit-test-1)
  - [Developing "Show - Wallet asynchronous event driven communication" (eventing)](#developing-show---wallet-asynchronous-event-driven-communication-eventing)
    - [`ChargeForReservation` Action](#chargeforreservation-action)
    - [Complete Reservation Action](#complete-reservation-action)
  - [End to end integration test](#end-to-end-integration-test)
    - [Run integration test](#run-integration-test-2)
- [Deploying](#deploying)
  - [Test service in production](#test-service-in-production)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
```

- The Wallet's `WalletCharged` domain event triggers the Show's `ConfirmPaymentConfirmation` command
- The Wallet's `WalletChargeRejected` domain event triggers the Show's `CancelSeatReservation` command

## Non-functional requirements
- effective asynchronous event driven communication
    - a common pattern to facilitate asynchronous, event-driven communication is to use a message broker as a data plane between bounded contexts, but it introduces another moving part
- strong delivery semantics
    - cannot lose any events
    - at-least-once delivery is required
- idempotency
    - deduplication with at-least-once delivery
- event ordering
    - ensure consistency
- design for failure
    - business or technical
- scalability
    - to be able to scale based on the demand

## Kalix components
[Kalix components](https://docs.kalix.io/concepts/programming-model.html#_kalix_components) abstract Kalix functionalities in high level building blocks.
Designing Kalix services is an exercise of combining and connecting Kalix components using domain driven design patterns.

### Show
![Show](images/KalixDesign-Show.png)

> For designing the "Show" bounded context in Kalix we are goning to use these Kalix components:

Reviewer: ```suggestion
For designing the "Show" bounded context in Kalix we are going to use these Kalix components:
```

The main advantages of a Kalix Entity over a Kalix View is providing an in-memory read view access in opposite to a Kalix View that is not kept in memory. A Kalix Entity can be used as a read view only if the read is view is queryable via a key/id.

In case of the "Show by reservation" view, a Kalix Entity is a suitable solution where `ReservationEntity` is modeled around a reservation and access via `reservationId` possible, it stores a reference to "Show" as `showId`. 
To be able to create instances of `ReservationEntity`, a "Fold show events to reservation" [Kalix Action](https://docs.kalix.io/java/actions.html) component is required to facilitate `subscribe` to `ShowEntity` events and to send `Create` command to `ReservationEntity`.

### Wallet

![Wallet](images/KalixDesign-Wallet.png)

For designing the Wallet bounded context in Kalix we are going to use for implementing Wallet (Entity/Aggregate) - [Kalix Entity (Event sourced)](https://docs.kalix.io/java/event-sourced-entities.html).

> **Note:** More information around implementing these Klaix service will be covered in following sections

Reviewer: ```suggestion
**Note:** More information around implementing these Kalix service will be covered in following sections
```

    if (currentState() != null) {
      return effects().error("show already exists", BAD_REQUEST);
    } else {
      return Show.ShowCreator.create(id, createShow).fold(
        error -> errorEffect(error, createShow),
        showCreated -> persistEffect(showCreated, "show created")
      );
    }
}
```

Via the `currentState()` helper method, Kalix gives us access to the current state.

Kalix ensures that each command handler endpoint, on a level of one instance of a Kalix Entity, is called in sequence, ensuring a sequential and consistent way of changing state.

`ShowCreated` event handler implementation:
```java
@EventHandler
public Show onEvent(ShowCreated showCreated) {
    return Show.create(showCreated);
}
```

**Note:** The `ShowEvent` sealed interface will give us a compile-time check that all event handlers are implemented.<br>
**Tip:** Check the `com.example.cinema.ShowEntity` class for a complete reference implementation.<br>

## Show Entity unit test
More information about Kalix Event source entity testing can be found [here](https://docs.kalix.io/java-protobuf/event-sourced-entities.html#_testing_the_entity).

> For testing the Kalix Entity, the Kalix Testkit is used (`EventSourcedTestKit`).

Reviewer: ```suggestion
For testing the Kalix Entity, the Kalix TestKit is used (`EventSourcedTestKit`).
```

  private final ComponentClient componentClient;

  public FoldShowEventsToReservationAction(ComponentClient componentClient) {
    this.componentClient = componentClient;
  }

  public Effect<String> onEvent(SeatReserved reserved) {
    return effects().forward(createReservation(reserved));
  }

  public Effect<String> onEvent(SeatReservationPaid paid) {
    return effects().forward(deleteReservation(paid.reservationId()));
  }

  private DeferredCall<Any, String> createReservation(SeatReserved reserved) {
    return componentClient.forValueEntity(reserved.reservationId())
      .call(ReservationEntity::create)
      .params(new ReservationEntity.CreateReservation(reserved.showId(), reserved.walletId(), reserved.price()));
  }

  private DeferredCall<Any, String> deleteReservation(String reservationId) {
    return componentClient.forValueEntity(reservationId).call(ReservationEntity::delete);
  }
}
```

Similar as with Kalix Views, the Action class level annotation `@Subscribe.EventSourcedEntity` defines a source. 
It configures an Action subscription to domain event changes from a Kalix Entity (`value = ShowEntity.class` indicates which Entity type).

> For each Kalix Entity domain event, the Action is subscribed to, we need to implement ahandler method.

Reviewer: ```suggestion
For each Kalix Entity domain event, the Action is subscribed to, we need to implement a handler method.
```

```
Define value for property 'groupId': `com.example`<br>
Define value for property 'artifactId': `java-spring-cinema-booking-choreography`<br>
Define value for property 'version' 1.0-SNAPSHOT: :<br>
Define value for property 'package' com.example: : `com.example`<br>
```

## Import the generated project into your IDE/editor

The generated Maven project has all required Maven plugins and dependencies needed to develop, test, run and deploy Kalix services.

## Adding additional, non-mandatory dependencies

Additional dependencies that are NOT mandatory for developing Kalix service but are convenient to be used in different situations (more explained in following sections).<br>
Add following dependencies in pom.xml:
```
<dependency>
  <groupId>io.vavr</groupId>
  <artifactId>vavr</artifactId>
  <version>0.10.4</version>
</dependency>
<dependency>
  <groupId>org.assertj</groupId>
  <artifactId>assertj-core</artifactId>
  <version>3.23.1</version>
  <scope>test</scope>
</dependency>
<dependency>
  <groupId>org.awaitility</groupId>
  <artifactId>awaitility</artifactId>
  <version>4.2.0</version>
</dependency>
```

`vavr` library is used to provide support for:
- immutable collections

> - Either support 

Reviewer: I've been considering whether vavr is a good fit for our use.

It seems suitable for the intermediate to advanced samples.

However, for the introductory level ones targeted at new Kalix users unfamiliar with vavr, it might be a distraction. Learning another new thing could divert their attention from the core Kalix concepts.

What do others think?

Reviewer 2: I had it in the user register choreography, but removed because it was not adding much value. Here is different. Logic is inside the domain, and the domain can reject a command by returning a Left with a message. 

The entity wrapper uses it to build a response and return an error with the message in passed with Left. 

But we could simplify it as well by having a more straightforward (and explicit) type like: `Result` that either has a list of events or an error message. Left / Right is not really expressive. 

By the way, we do have a similar a class already. It's called `Response`. The one that I suggested to move to `com.example.common`. 

That said, if we are adding vavr just because of Either, then we can simply build our own more expressive variation of it.

My 2cc.

Author: we use `vavr` also for immutable map

Reviewer 2: Correct. And that's an important topic (and old topic). Mutable objects (including collections) is a problem. Using record is not enough because a record can reference mutable objects themselves.

If someone uses a mutable data structure and mutates it inside the command handler, the state will be effectively mutated and eventually snapshotted without the user being aware of it.

Reviewer: [Today I learned](https://typesafe.slack.com/archives/C02A6LVUZ/p1710934023470989) the future of the Vavr project appears to be uncertain https://danieldietrich.dev/blog/2023/12/13/drawing-a-clear-line/.

Due to this uncertainty, we should consider removing its usage from this sample.
